### PR TITLE
Take over: New lint bytes count to len

### LINKED ISSUE #8083: Lint suggesting to replace `str::bytes().count()` with `str::len()`
### What it does

It checks for `str::bytes().count()` and suggests replacing it with `str::len()` that does the same but shorter and possibly more performant.

Extra: can something like this be done for https://doc.rust-lang.org/std/io/trait.Read.html#method.bytes too?

### Lint Name

bytes_count_to_len

### Category

complexity

### Advantage

- Might be more performant
- Shorter
- Possibly easier to read

### Drawbacks

_No response_

### Example

```rust
fn main() {
    "hello".bytes().count();
}
```

Could be written as:

```rust
fn main() {
    "hello".len();
}
```

### PR DESCRIPTION
**Pull request by kyoto7250**
take over #8375 
close #8083 

This PR adds new lint about  considering replacing `.bytes().count()` with `.len()`.

Thank you in advance.

---

r! @Manishearth 

changelog: adds new lint [`bytes_count_to_len`] to consider replacing `.bytes().count()` with `.len()`

### CODE REVIEW

File: clippy_lints/src/bytes_count_to_len.rs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+            if is_type_diagnostic_item(cx, ty, sym::String) || ty.kind() == &ty::Str;
+            then {
+                let mut applicability = Applicability::MachineApplicable;
+                span_lint_and_sugg(
+                    cx,
+                    BYTES_COUNT_TO_LEN,
+                    expr.span,
+                    "using long and hard to read `.bytes().count()`",
+                    "consider calling `.len` instead:",
```
Comment (matthiaskrgr) [inline clippy_lints/src/bytes_count_to_len.rs:60]: "consider calling `.len()` instead` (you can skip the ":" here I think, looks like it is already inserted in the warning message automatically? (see tests output)

Comment (kyoto7250) [reply clippy_lints/src/bytes_count_to_len.rs:60]: That's right.

Thanks!

**Comment by giraffate** [comment]
Could you do a rebase, not merge? We follow a no merge-commit policy: https://github.com/rust-lang/rust-clippy/blob/master/doc/basics.md#pr

File: clippy_lints/src/bytes_count_to_len.rs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// "hello".len();
+    /// String::from("hello").len();
+    /// ```
+    #[clippy::version = "1.62.0"]
+    pub BYTES_COUNT_TO_LEN,
+    complexity,
+    "Using bytes().count() when len() performs the same functionality"
```
Comment (giraffate) [inline clippy_lints/src/bytes_count_to_len.rs:34]: nits
```suggestion
    "Using `bytes().count()` when `len()` performs the same functionality"
```

File: clippy_lints/src/bytes_count_to_len.rs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+            if is_type_diagnostic_item(cx, ty, sym::String) || ty.kind() == &ty::Str;
+            then {
+                let mut applicability = Applicability::MachineApplicable;
+                span_lint_and_sugg(
+                    cx,
+                    BYTES_COUNT_TO_LEN,
+                    expr.span,
+                    "using long and hard to read `.bytes().count()`",
+                    "consider calling `.len` instead",
```
Comment (giraffate) [inline clippy_lints/src/bytes_count_to_len.rs:60]: nits
```suggestion
                    "consider calling `.len()` instead",
```

File: clippy_lints/src/bytes_count_to_len.rs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    "Using bytes().count() when len() performs the same functionality"
+}
+
+declare_lint_pass!(BytesCountToLen => [BYTES_COUNT_TO_LEN]);
+
+impl<'tcx> LateLintPass<'tcx> for BytesCountToLen {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'_>) {
+        if_chain! {
+            if let hir::ExprKind::MethodCall(expr_method, expr_args, _) = &expr.kind;
+            if expr_method.ident.name == rustc_span::sym::count;
```
Comment (giraffate) [inline clippy_lints/src/bytes_count_to_len.rs:43]:
```suggestion
            if expr_method.ident.name == sym::count;
```

**Comment by kyoto7250** [comment]
> Could you do a rebase, not merge? We follow a no merge-commit policy: https://github.com/rust-lang/rust-clippy/blob/master/doc/basics.md#pr

Thanks, I missed the description.

I will rebase this PR, and fix it.

**Comment by kyoto7250** [comment]
I have corrected my PR and would like to get another review.

**Review by giraffate — COMMENTED** [review]
Can this review https://github.com/rust-lang/rust-clippy/pull/8375#discussion_r798200109 be addressed?

**Comment by kyoto7250** [comment]
@giraffate 

I have addressed [this commit](https://github.com/rust-lang/rust-clippy/pull/8711/commits/38bcd8ebc1252b1ed9a9b280510fa57b46ac5235) in this commit, but I think we need to add a test that addresses this fix.

How do I add a test about this case?

**Comment by giraffate** [comment]
https://github.com/rust-lang/rust-clippy/pull/8711#issuecomment-1101844729

Although `str` has already ensured in https://github.com/rust-lang/rust-clippy/blob/26732f121c9782d95c1e05bd825f447833730fe8/clippy_lints/src/bytes_count_to_len.rs#L55, we add the test case for it like this:
```rs
struct Foo<'a> {
    str: &'a str,
}

impl Foo<'_> {
    fn bytes(&self) -> std::str::Bytes<'_> {
        self.str.bytes()
    }
}

fn main() {
    let f = Foo { str: "foo" };
    let _ = f.bytes().count();
}
```

**Review by giraffate — APPROVED** [review]
Overall looks good. Could you squash commits?

**Comment by kyoto7250** [comment]
@giraffate 

I made two commits with c-rus commit and my commit.

**Comment by giraffate** [comment]
@bors r+

Thanks!

**Pull request merged**